tensorboard/plugins/graph/tf_graph_common/op.ts

```diff
@@ -14,435 +14,438 @@ limitations under the License.
+      // compatibility for various reasons.
 
-  /**
-   * Returns true if OpNode graph object represents a
-   * Tensorflow operation that is valid for the TPU.
-   */
-  export function opValid(opNode: OpNode): boolean {
-    // Function library nodes are generally for internal use.
-    if (opNode.name.search(FUNCTION_LIBRARY_NODE_PREFIX) == 0) {
-      return true;
-    }
-    // Nodes that lack op types should be ignored.
-    if (!opNode.op) {
-      return true;
-    }
-    // If assigned a device that is not TPU-related assume op is valid.
-    if (opNode.device && isNotTpuOp(opNode.device)) {
-      return true;
+      // Control flow ops.
+      'Enter',
+      'Exit',
+      'LoopCond',
+      'Merge',
+      'NextIteration',
+      'Switch',
+      // Ops below are inserted by the compiler.
+      '_Arg',
+      '_ParallelConcatUpdate',
+      '_Retval',
+      '_TPUCompile',
+      '_TPUExecute',
+      // Distributed TPU ops.
+      'TPUCompilationResult',
+      'TPUReplicatedInput',
+      'TPUReplicatedOutput',
+      'TPUReplicateMetadata',
+      // Checkpointing ops.
+      'MergeV2Checkpoints',
+      'RestoreV2',
+      'SaveV2',
+      // Miscellaneous CPU ops.
+      'Abort',
+      'Assert',
+      'Assign',
+      'Placeholder',
+      'PlaceholderV2',
+      'ShardedFilename',
+      'StringJoin',
+      'Variable',
+      'VariableV2',
+      'VarHandleOp',
+      // Summary ops.
+      'AudioSummary',
+      'AudioSummaryV2',
+      'DebugNumericSummary',
+      'HistogramSummary',
+      'ImageSummary',
+      'MergeSummary',
+      'ScalarSummary',
+      'StatsAggregatorSummary',
+    ];
+
+    /**
+     * Returns true if the node's inferred device is not the TPU.
+     * Note that this is only a best-effort check.
+     */
+    isNotTpuOp(opDevice: string): boolean {
+      if (opDevice.toLowerCase().search('cpu:') != -1) {
+        return true;
+      }
+      if (opDevice.toLowerCase().search('gpu:') != -1) {
+        return true;
+      }
+      return (opDevice.toLowerCase().search('tpu') == -1);
     }
-    // If assigned to the TPU_SYSTEM device, assume op is valid.
-    if (opNode.device && opNode.device.search('TPU_SYSTEM') != -1) {
-      return true;
+    opValid(opNode: OpNode): boolean {
+      // Function library nodes are generally for internal use.
+      if (opNode.name.search(FUNCTION_LIBRARY_NODE_PREFIX) == 0) {
+        return true;
+      }
+      // Nodes that lack op types should be ignored.
+      if (!opNode.op) {
+        return true;
+      }
+      // If assigned a device that is not TPU-related assume op is valid.
+      if (opNode.device && this.isNotTpuOp(opNode.device)) {
+        return true;
+      }
+      // If assigned to the TPU_SYSTEM device, assume op is valid.
+      if (opNode.device && opNode.device.search('TPU_SYSTEM') != -1) {
+        return true;
+      }
+      return TpuCompatibilityProvider.WHITELIST.indexOf(opNode.op) != -1;
```

Comment:
`Array.prototype.includes` is more semantically correct than `Array.prototype.indexOf`. Please prefer `includes`.

Reply:
The `Array.prootype.includes` is defined in ES2016(ES7), but the tensorboard BUILD is using es6, so we can't use `includes` directly. 
See also:
https://github.com/tensorflow/tensorboard/blob/d285c7de73a09e9ece5a13368a58c00e1dcabda9/tensorboard/plugins/graph/tf_graph_common/op.ts#L432

Reply:
I suppose `_.includes(collection, value)` from lodash lib.